cpu/stm32/periph/adc_l1.c

```diff
@@ -157,12 +155,17 @@ int32_t adc_sample(adc_t line, adc_res_t res)
     /* lock and power on the ADC device  */
     prep();
 
-    /* set resolution, conversion channel and single read */
-    ADC1->CR1 |= res & ADC_CR1_RES;
+    /* mask and set resolution, conversion channel and single read */
+    ADC1->CR1 = (ADC1->CR1 & ~ADC_CR1_RES_Msk) | (res & ADC_CR1_RES);
```

Comment:
something seems wrong to me to my with this application of res values

```
ADC1->CR1 = (ADC1->CR1 & ~ADC_CR1_RES_Msk) | ((res << ADC_CR1_RES_Pos) & ADC_CR1_RES_Msk );
```
would be what i expect but on the other hand  the values in `adc_res_t` are shifted 
```
 typedef enum {
    ADC_RES_6BIT  = (0x3 << 3),     /**< ADC resolution: 6 bit */
    ADC_RES_8BIT  = (0x2 << 3),     /**< ADC resolution: 8 bit */
    ADC_RES_10BIT = (0x1 << 3),     /**< ADC resolution: 10 bit */
    ADC_RES_12BIT = (0x0 << 3),     /**< ADC resolution: 12 bit */
    ADC_RES_14BIT = (0xfe),         /**< not applicable */
    ADC_RES_16BIT = (0xff)          /**< not applicable */
} adc_res_t;
```

but not by the correct ammount:

```
cmsis/l1/Include/stm32l152xd.h
999:#define ADC_CR1_RES_Pos                      (24U)                             
1000:#define ADC_CR1_RES_Msk                      (0x3UL << ADC_CR1_RES_Pos)         /*!< 0x03000000 */
1001:#define ADC_CR1_RES                          ADC_CR1_RES_Msk                   /*!< ADC resolution */
1002:#define ADC_CR1_RES_0                        (0x1UL << ADC_CR1_RES_Pos)         /*!< 0x01000000 */
1003:#define ADC_CR1_RES_1                        (0x2UL << ADC_CR1_RES_Pos)         /*!< 0x02000000 */
```

Reply:
would you please take another look at that

Reply:
I think you looked into the header of the STM32L0, which has a different position of the resolution bits:

`cpu/stm32/include/periph/l0/periph_cpu.h` file: https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32/include/periph/l0/periph_cpu.h#L41-L51
```
#define HAVE_ADC_RES_T
typedef enum {
    ADC_RES_6BIT  = (0x3 << 3),     /**< ADC resolution: 6 bit */
    ADC_RES_8BIT  = (0x2 << 3),     /**< ADC resolution: 8 bit */
    ADC_RES_10BIT = (0x1 << 3),     /**< ADC resolution: 10 bit */
    ADC_RES_12BIT = (0x0 << 3),     /**< ADC resolution: 12 bit */
    ADC_RES_14BIT = (0xfe),         /**< not applicable */
    ADC_RES_16BIT = (0xff)          /**< not applicable */
} adc_res_t;
/** @} */
#endif /* ndef DOXYGEN */
```

The L1 already has the improved style for the resolution:

`cpu/stm32/include/periph/l1/periph_cpu.h` file: https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32/include/periph/l1/periph_cpu.h#L42-L56
```cpp
/**
 * @brief   Override the ADC resolution configuration
 * @{
 */
#define HAVE_ADC_RES_T
typedef enum {
    ADC_RES_6BIT  = (ADC_CR1_RES_0 | ADC_CR1_RES_1),    /**< ADC resolution: 6 bit */
    ADC_RES_8BIT  = (ADC_CR1_RES_1),                    /**< ADC resolution: 8 bit */
    ADC_RES_10BIT = (ADC_CR1_RES_0),                    /**< ADC resolution: 10 bit */
    ADC_RES_12BIT = (0x00),                             /**< ADC resolution: 12 bit */
    ADC_RES_14BIT = (0xfe),                             /**< not applicable */
    ADC_RES_16BIT = (0xff)                              /**< not applicable */
} adc_res_t;
/** @} */
#endif /* ndef DOXYGEN */
```

The resolution flag position (and registers) vary wildly between different STM32 families, and the implementation style in RIOT varies just as much, which was the motivation for #20971.

Reply:
These are the relevant pages in the reference manuals. The register positions for both variants are correct.

https://www.st.com/resource/en/reference_manual/rm0367-ultralowpower-stm32l0x3-advanced-armbased-32bit-mcus-stmicroelectronics.pdf page 341

https://www.st.com/resource/en/reference_manual/cd00240193-stm32l100xx-stm32l151xx-stm32l152xx-and-stm32l162xx-advanced-arm-based-32-bit-mcus-stmicroelectronics.pdf page 291